cohortextractor/tpp_backend.py

```diff
@@ -42,14 +42,16 @@ def to_csv(self, filename):
             cleanup_queries = []
         temp_filename = self._get_temp_filename(filename)
         unique_check = UniqueCheck()
-        result = self.execute_queries(queries)
-        with open(temp_filename, "w", newline="") as csvfile:
-            writer = csv.writer(csvfile)
-            writer.writerow([x[0] for x in result.description])
-            for row in result:
-                writer.writerow(row)
-                # The first column contains IDs
-                unique_check.add(row[0])
+        result_cursor = self.execute_queries(queries)
+        # `batch_size` here was chosen through a bit of unscientific
+        # trial-and-error and some guesswork. It may well need changing in
+        # future.
+        rows = dbapi_cursor_to_csv_file(
+            result_cursor, temp_filename, batch_size=32000, yield_output_rows=True
+        )
+        for row in rows:
+            # The first column contains IDs
+            unique_check.add(row[0])
```

Comment:
Would it be conceptually simpler to produce the CSV files first, and then do the uniqueness check?